mathesar/api/db/permissions/table.py

```diff
@@ -56,7 +56,24 @@ def _scope_queryset(cls, request, qs, allowed_roles):
             permissible_schema_roles_filter = (
                 Q(schema__schema_role__role__in=allowed_roles) & Q(schema__schema_role__user=request.user)
             )
+            permissible_schema_role_filter = Q(schema__schema_role__role__in=confirmed_table_roles)
+            permissible_table_view_filter = Q(import_verified=True) | Q(import_verified__isnull=True)
+            confirmed_table = "CONFIRMED"
+            unconfirmed_table = "UNCONFIRMED"
+            not_allowed = "NOT_ALLOWED"
+
             qs = qs.filter(permissible_database_role_filter | permissible_schema_roles_filter)
+
+            qs = qs.annotate(
+                to_count=Case(
+                    When(permissible_schema_role_filter, then=Case(
+                        When(permissible_table_view_filter, then=Value(confirmed_table)),
+                        default=Value(unconfirmed_table)
+                    )),
+                    default=Value(not_allowed), output_field=CharField()
+                )
+            )
+            qs = qs.filter(Q(to_count=not_allowed) | Q(to_count=confirmed_table))
```

Comment:
This whole logic can be boiled down to 
```
(Q(schema__schema_role__role='manager') & Q(schema__schema_role__user=request.user) & import_verified=False)) | (permissible_schema_roles_filter & (Q(import_verified=True) | Q(import_verified__is_null=True)))
```
The first part before the `|` (OR) returns all the tables with import_verified set to false if the user is a manager and the later part will return tables with import_verified set to true and null.

Reply:
Oh yeah, this is a way more concise and straightforward solution. I had some doubts about the database_role query though. So as I understand things `/api/db/v0/tables/` needs to return tables from other databases as well. From the workflow that I could understand if the `allowed_roles` for the database is different (in a hypothetical scenario) then it makes sense to only show confirmed & unconfirmed tables or only confirmed tables (depending upon the schema role) of that specific database. 

For example, if my user has the role of Manager and Viewer in the database with ids 1 and 2 respectively, and my allowed roles for permissible_database_role_filter are Manager and Editor. In this case, I only need to show schemas from database_1 and not database_2 right?
If this is the case we can extend over the above query and just make an AND operation to filter out the required tables.
```
permissible_database_role_filter = (
                Q(schema__database__database_role__role__in=allowed_roles)
                & Q(schema__database__database_role__user=request.user)
            )
permissible_schema_roles_filter = (
    Q(schema__schema_role__role__in=allowed_roles) & Q(schema__schema_role__user=request.user)
)
cnf_table_schema_role_access_filter = (Q(schema__schema_role__role='manager') & Q(schema__schema_role__user=request.user) & Q(import_verified=False))
cnf_table_filter = (permissible_schema_roles_filter & (Q(import_verified=True) | Q(import_verified__isnull=True)))
qs = qs.filter((cnf_table_filter | cnf_table_schema_role_access_filter) & permissible_database_role_filter)
```

Reply:
> We treat database_roles differently than schema_roles and database_roles should be treated hierarchially greater than any schema_roles. The query suggested by you will result in a null set if the user doesn't have any database_roles inspite of having valid schema_roels as you are doing an AND operation.

So if I am interpreting this correctly, database_roles are supposed to take precedence in case there is a conflict between database_roles and schema_roles? And if a user does not have a database_role it defaults to the permissions of schema_role?

I researched a bit about these user roles in the GitHub issues and documentation and found #2524 which says
> * On the backend, if the user has 2 permission access levels, one on the database and one on the schema, the highest permission level will take precedence.
>    
>   * If user is `viewer` on db, and `manager` on schema, the user will have `manager` access on the schema.
>   * If user is `manager` on db, and `editor` on schema, the user will have `manager` access on the schema.
> * The frontend behaviour is currently: access level on schema takes precedence over access level on the database. By this definition,
>   
>   * If user is `viewer` on db, and `manager` on schema, the user will have `manager` access on the schema.
>   * If user is `manager` on db, and `editor` on schema, the user will have `editor` access on the schema.

If I have understood your reply correctly, the flow, in this case, should've been this; as the database_role is dominating over schema_role.
* If user is `viewer` on db, and `manager` on schema, the user will have `user` access on the schema.
* If user is `manager` on db, and `editor` on schema, the user will have `manager` access on the schema.

Doesn't this go against the flow being discussed in that issue?
If I follow the rule of highest role precedence, the query can be modified into this:
```
cnf_table_schema_role_access_filter = (Q(schema__schema_role__role='manager') & Q(schema__schema_role__user=request.user) & Q(import_verified=False))
cnf_table_database_role_access_filter = (Q(schema__database__database_role__role='manager') & Q(schema__database__database_role__user=request.user) & Q(import_verified=False))
cnf_table_filter = (permissible_schema_roles_filter & (Q(import_verified=True) | Q(import_verified__isnull=True)))
qs = qs.filter(cnf_table_filter | cnf_table_schema_role_access_filter | cnf_table_database_role_access_filter).distinct()

```

I do think that there is a gap in my understanding with respect to what the database_role is meant to do. Reading up the [docs](https://wiki.mathesar.org/en/product/specs/users-permissions) for user permissions, I find that schema permissions involve adding and removing tables. Now since the confirmed/unconfirmed tables feature is closely tied in with the addition of tables why does the database_role matter here? So going back to our earlier query:

```
cnf_table_filter = (permissible_schema_roles_filter & (Q(import_verified=True) | Q(import_verified__isnull=True)))
cnf_table_schema_role_access_filter = (Q(schema__schema_role__role='manager') & Q(schema__schema_role__user=request.user) & Q(import_verified=False))
qs = qs.filter(cnf_table_filter | cnf_table_schema_role_access_filter).distinct()
```
> The query suggested by you will result in a null set if the user doesn't have any database_roles inspite of having valid schema_roels as you are doing an AND operation.

This query satisfies this problem. I know I am starting to look a bit silly now but what edge case are we trying to solve in this query that requires the use of database_roles?